### PR TITLE
CRM-19150 Preserve contribution when associated membership is deleted (step 1 of 2 | API and BAO)

### PR DESCRIPTION
**Pull request by joseltorres**
Currently when a membership is deleted, if there is an associated contribution then the contribution also gets deleted. I would like to see an option to keep the preserve the contrition. I suggest this be completed in two steps: 1) The ability the preserver the contribution through code (API and BAO) and 2) UI change to allow front-end users to check a box to preserve the associated contribution.

Added optional parameters in the the API and BAO that will unlink membership and contribution, then skip-over deleting contribution.

https://issues.civicrm.org/jira/browse/CRM-19150

### CODE REVIEW

**Comment by totten** [comment]
jenkins, add to whitelist

**Comment by seamuslee001** [comment]
@joseltorres Jose it appears there are some style issues to fix up here https://test.civicrm.org/job/CiviCRM-Core-PR/10914/checkstyleResult/new/

File: CRM/Member/BAO/Membership.php [hunk 1]
```diff
@@ -1500,14 +1500,16 @@ public static function createRelatedMemberships(&$params, &$dao, $reset = FALSE)
...
-  public static function deleteMembershipPayment($membershipId) {
+  public static function deleteMembershipPayment($membershipId, $preserveContrib = 0) {
 
     $membershipPayment = new CRM_Member_DAO_MembershipPayment();
     $membershipPayment->membership_id = $membershipId;
     $membershipPayment->find();
 
     while ($membershipPayment->fetch()) {
-      CRM_Contribute_BAO_Contribution::deleteContribution($membershipPayment->contribution_id);
+      if(!$preserveContrib) {
```
Comment (monishdeb) [inline CRM/Member/BAO/Membership.php]: please change ```if(``` to ```if (```

File: api/v3/Membership.php [hunk 2]
```diff
@@ -42,7 +54,12 @@
  *   API result array.
  */
 function civicrm_api3_membership_delete($params) {
-  return _civicrm_api3_basic_delete(_civicrm_api3_get_BAO(__FUNCTION__), $params);
+  if (CRM_Member_BAO_Membership::del($params['id'], $params['preserve_contribution'])) {
+    return civicrm_api3_create_success(true, $params);
```
Comment (monishdeb) [inline api/v3/Membership.php]: Change ```true``` to ```TRUE```

**Comment by monishdeb** [comment]
@joseltorres can you add a api test for this improvement. Also it seems there is a false merge in the commit list, so please create a separate PR by cherry-picking desired commits. Such false merge arise when you try to pull in your branch instead of merge. In order to rebase  you need to ```git pull --rebase upstream master```

**Comment by joseltorres** [comment]
creating a new PR, rebased

**Pull request closed without merging**